tests/algorithms/test_gradient_clipping.py

```diff
@@ -58,56 +58,122 @@ def forward(self, x):
 
-    model = simple_model_with_grads
-    # Make sure after calling apply_agc, the gradients inside the model are
-    # the same as if we manually called _get_clipped_gradients on the weights and
-    # gradients.
-    weights = next(model.parameters())
-    grad = weights.grad
-    expected_clipped_grad = grad.detach() * _get_clipped_gradient_coeff(weights, grad)
-    cf.apply_agc(model)
-    current_grad = next(model.parameters()).grad
-    torch.equal(current_grad, expected_clipped_grad)
+    new_gc_fn = Mock()
+    monkeypatch.setattr(torch.nn.utils, 'clip_grad_norm_', new_gc_fn)
+    apply_gradient_clipping(parameters, 'norm', 0.1)
+    new_gc_fn.assert_called_once()
 
+    new_gc_fn = Mock()
+    monkeypatch.setattr(torch.nn.utils, 'clip_grad_value_', new_gc_fn)
+    apply_gradient_clipping(parameters, 'value', 0.1)
+    new_gc_fn.assert_called_once()
 
-def test_agc_functional_with_cnn_does_not_error(cnn_model_with_grads):
-    """This test is just to ensure that no errors are raised.
 
-    Accuracy of the AGC calculations are tested in other tests.
-    """
+@pytest.mark.parametrize('clipping_type', [('adaptive',), ('norm',), ('value',)])
+def test_gradient_clipping_algorithm(monkeypatch, clipping_type, simple_model_with_grads):
+    model = simple_model_with_grads
+    apply_gc_fn = Mock()
+    monkeypatch.setattr(gc_module, 'apply_gradient_clipping', apply_gc_fn)
+    state = Mock()
+    state.is_model_deepspeed = False
+    state.model = model
+    state.profiler.marker = Mock(return_value=None)
+    state.callbacks = []
+    state.algorithms = [GradientClipping(clipping_type=clipping_type, clipping_threshold=0.01)]
+    logger = Mock()
+    engine = Engine(state, logger)
 
-    model = cnn_model_with_grads
-    # Call apply_agc. If this function returns then we know that nothing errored out.
-    cf.apply_agc(model)
+    # Run the Event that should cause AGC.apply to be called.
+    engine.run_event(Event.AFTER_TRAIN_BATCH)
 
+    apply_gc_fn.assert_called_once()
 
-def test_AGC_algorithm(simple_model_with_grads):
-    # Get weight and gradients from the fixture model.
-    model = simple_model_with_grads
-    weights = next(model.parameters())
-    grad = weights.grad
-    expected_clipped_grad = grad.detach() * _get_clipped_gradient_coeff(weights, grad)
 
-    # Set up a mock engine.
+def test_gradient_clipping_algorithm_with_deepspeed_enabled(monkeypatch: pytest.MonkeyPatch):
+    clipping_threshold = 0.1191
+    apply_gc_fn = Mock()
+    monkeypatch.setattr(gc_module, 'apply_gradient_clipping', apply_gc_fn)
     state = Mock()
-    state.model = model
     state.profiler.marker = Mock(return_value=None)
     state.callbacks = []
-    state.algorithms = [AGC()]
+
+    # Enable deepspeed and set clipping_type to norm to ensure that apply_gradient_clipping
+    # is not called.
+    state.algorithms = [GradientClipping(clipping_type='norm', clipping_threshold=clipping_threshold)]
+    state.is_model_deepspeed = True
```

Comment:
This is a read-only property so it cannot be set like this. Intead, the model will have to be wrapped with DeepSpeed engine. It would probably be easiest to construct an actual trainer here that uses deepspeed.

Reply:
fixed here: [cc8830d](https://github.com/mosaicml/composer/pull/1099/commits/cc8830dc422f9a472fc61fba673f07ba9e2d22b3). Using `state.deepspeed_enabled` which checks `state.deepspeed_config`